projs/shadow/shadow-engine/core/inc/event-bus/event_bus.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <string>
+#include <functional>
+#include <map>
+#include <memory>
+#include <any>
+
+#include "SHObject.h"
+#include "events.h"
+
+namespace ShadowEngine::EventBus {
+
+    template<class T, class U>
+    concept Derived = std::is_base_of<T, U>::value;
+
+    template<typename T>
+    concept SHObjectConcept = Derived<ShadowEngine::SHObject, T>;
+
+
+
+
+    template<typename T>
+    concept EventType = Derived<Event, T>;
+
+    template<EventType T>
+    struct Subscription {
+        using Callback = std::function<const void(T &)>;
+
+        Callback callback;
+
+        std::weak_ptr<ShadowEngine::SHObject> binding;
+
+        explicit Subscription(Callback call): callback(call){}
+    };
+
+    class EventDispatcherBase {
+    public:
+        virtual ~EventDispatcherBase() = default;
+
+        virtual void call(ShadowEngine::SHObject &obj) = 0;
+    };
+
+    template<EventType T>
+    class EventDispatcher : public EventDispatcherBase {
+    public:
+        using SubRef = std::shared_ptr<Subscription<T>>;
+
+    protected:
+        using SubscriptionList = std::vector<SubRef>;
+
+        SubscriptionList subscriptionList;
+
+    public:
+        std::shared_ptr<Subscription<T>> subscribe(const std::shared_ptr<ShadowEngine::SHObject> &binding, std::function<const void(T &)> func) {
```

Comment:
Binding is a remnant of an older idea where events would automatically unsubscribe based on a weak(?) pointer to the owner. This was abandoned because making sure a managed pointer was available at the point of the subscription was hard.

If you have any ideas for this kind of "owner" based subscription please add a comment about it